File: scripts/analyze-scenes-151-200.py
#!/usr/bin/env python3
import json
import re
from pathlib import Path
from typing import List, Dict, Set, Any

def extract_all_scenes(outline_data: Dict) -> List[Dict]:
    """Extract all scenes from the outline in order."""
    all_scenes = []
    global_index = 0

    # Navigate through the structure
    books_data = outline_data.get('SelfImprovementSeries', {}).get('Books', {}).get('trilogies', {})

    # Process each trilogy
    for trilogy_key in sorted(books_data.keys()):
        trilogy = books_data[trilogy_key]
        trilogy_books = trilogy.get('trilogy_books', {})

        # Process each book in the trilogy
        for book_key in sorted(trilogy_books.keys()):
            book = trilogy_books[book_key]
            task_masters = book.get('task_masters', {})

            # Process each task master
            for tm_key in sorted(task_masters.keys()):
                task_master = task_masters[tm_key]
                major_task_groups = task_master.get('major_task_groups', {})

                # Process each major task group
                for mtg_key in sorted(major_task_groups.keys()):
                    major_task_group = major_task_groups[mtg_key]

                    # Check if this is a major_activity_theme (different structure)
                    if 'major_activity_theme' in mtg_key:
                        continue

                    specific_task_groups = major_task_group.get('Specific_task_groups', {})

                    # Process each specific task group (chapter)
                    for stg_key in sorted(specific_task_groups.keys()):
                        specific_task_group = specific_task_groups[stg_key]
                        chapter_id = specific_task_group.get('unique_identifier', '')
                        scenes = specific_task_group.get('scenes', [])

                        # Process each scene
                        for scene in scenes:
                            global_index += 1
                            scene_data = scene.copy()
                            scene_data['global_index'] = global_index
                            scene_data['chapter_id'] = chapter_id
                            scene_data['chapter_title'] = specific_task_group.get('chapter_title', '')
                            all_scenes.append(scene_data)

    return all_scenes

def extract_characters(scene: Dict) -> List[str]:
    """Extract character names from scene data."""
    characters = set()

    # Add POV character
    if scene.get('pov'):
        characters.add(scene['pov'])

    # Add from characters array if exists
    if scene.get('characters') and isinstance(scene['characters'], list):
        characters.update(scene['characters'])

    # Known characters to search for
    known_characters = [
        'Francisco', 'Zara', 'Roger', 'Maggie', 'Marcus', 'Kai', 'Elena',
        'Thomas', 'Sarah', 'David', 'Michael', 'Lisa', 'James', 'Amara',
        'Viktor', 'Isabella', 'Chen', 'Olivia', 'Nathan', 'Emma', 'Alex',
        'Sophie', 'Daniel', 'Rachel', 'Liam', 'Grace', 'Ethan', 'Maya',
        'Jordan', 'Chloe', 'Ryan', 'Ava', 'Noah', 'Mia', 'Lucas', 'Aria',
        'Dr. Martinez', 'Dr. Chen', 'Dr. Patel', 'Dr. Kim', 'Professor',
        'Coach', 'Mentor', 'Instructor', 'Director', 'Manager', 'CEO'
    ]

    # Search in text fields
    text_fields = ' '.join(filter(None, [
        scene.get('setup', ''),
        scene.get('description', ''),
        scene.get('focus', ''),
        scene.get('scene_title', ''),
        scene.get('title', ''),
        scene.get('internal_conflict', ''),
        scene.get('internalConflict', ''),
        scene.get('preliminary_scene_focus', ''),
        scene.get('preliminary_scene_description', ''),
        str(scene.get('character_arcs', '')),
        str(scene.get('interactions', ''))
    ]))

    for name in known_characters:
        if name in text_fields:
            characters.add(name)

    return sorted(list(characters))

def enhance_internal_conflict(scene: Dict, characters: List[str]) -> str:
    """Create enhanced internal conflict that includes multi-character dynamics."""
    current_conflict = scene.get('internal_conflict') or scene.get('internalConflict', '')
    pov = scene.get('pov', 'Unknown POV')
    title = scene.get('scene_title') or scene.get('title', '')

    # If there's already a detailed conflict, check if it mentions other characters
    if current_conflict and len(current_conflict) > 100:
        has_other_characters = any(c != pov and c in current_conflict for c in characters)
        if has_other_characters:
            return current_conflict

    # Build enhanced conflict based on available information
    setup = scene.get('setup', '')
    description = scene.get('description', '')
    focus = scene.get('focus', '')
    preliminary_focus = scene.get('preliminary_scene_focus', '')
    preliminary_desc = scene.get('preliminary_scene_description', '')
    beat_goal = scene.get('beat_goal', '')
    core_emotion = scene.get('core_emotion', '')
    character_arcs = scene.get('character_arcs', '')

    enhanced = current_conflict or ''

    # Add context from other fields if conflict is minimal
    if len(enhanced) < 50:
        context_parts = []
        if beat_goal:
            context_parts.append(f"Goal: {beat_goal}")
        if core_emotion:
            context_parts.append(f"Emotion: {core_emotion}")
        if preliminary_focus:
            context_parts.append(preliminary_focus)
        if preliminary_desc:
            context_parts.append(preliminary_desc)
        if focus:
            context_parts.append(focus)
        if description:
            context_parts.append(description)
        if setup:
            context_parts.append(setup)
        if character_arcs:
            context_parts.append(f"Character arcs: {character_arcs}")

        if context_parts:
            enhanced = ' | '.join(context_parts)[:500]

    # Add character interaction context if multiple characters are present
    if len(characters) > 1:
        other_characters = [c for c in characters if c != pov]
        if other_characters and not any(oc in enhanced for oc in other_characters):
            enhanced += f" [Involves interactions with: {', '.join(other_characters)}]"

    # If still empty, create a placeholder
    if not enhanced:
        enhanced = f"[Scene {scene.get('global_index')}: {title}] - Internal conflict details needed. POV: {pov}"

    return enhanced

def main():
    # Read the outline file
    outline_path = Path(__file__).parent.parent / 'data' / 'l_outline.json'
    print(f"Reading outline from: {outline_path}")

    with open(outline_path, 'r', encoding='utf-8') as f:
        outline_data = json.load(f)

    # Extract all scenes
    print("Extracting all scenes...")
    all_scenes = extract_all_scenes(outline_data)
    print(f"Total scenes found: {len(all_scenes)}")

    # Extract scenes 151-200
    target_scenes = [s for s in all_scenes if 151 <= s['global_index'] <= 200]
    print(f"Scenes 151-200: {len(target_scenes)} scenes")

    if not target_scenes:
        print("No scenes found in range 151-200!")
        print(f"Available scene range: 1-{len(all_scenes)}")
        return

    # Analyze each scene
    analysis = []
    for scene in target_scenes:
        characters = extract_characters(scene)

        analysis_entry = {
            'scene_index': scene['global_index'],
            'chapter_id': scene['chapter_id'],
            'scene_number': scene.get('scene_number', 0),
            'title': scene.get('scene_title') or scene.get('title', 'Untitled'),
            'pov': scene.get('pov', 'Unknown'),
            'current_internal_conflict': scene.get('internal_conflict') or scene.get('internalConflict', ''),
            'characters_involved': characters,
            'proposed_enhanced_conflict': enhance_internal_conflict(scene, characters)
        }
        analysis.append(analysis_entry)

    # Write the analysis to output file
    output_path = Path(__file__).parent / 'scenes-151-200-analysis.json'
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(analysis, f, indent=2, ensure_ascii=False)

    print(f"\nAnalysis complete!")
    print(f"Output written to: {output_path}")
    print(f"\nSummary:")
    print(f"- Total scenes analyzed: {len(analysis)}")
    print(f"- Scenes with internal conflicts: {sum(1 for a in analysis if a['current_internal_conflict'])}")
    print(f"- Scenes with multiple characters: {sum(1 for a in analysis if len(a['characters_involved']) > 1)}")

    # Character distribution
    all_chars = set()
    for a in analysis:
        all_chars.update(a['characters_involved'])
    print(f"- Unique characters involved: {len(all_chars)}")
    print(f"- Characters: {', '.join(sorted(all_chars))}")

    # Print sample
    print(f"\nSample scenes:")
    for a in analysis[:5]:
        print(f"\nScene {a['scene_index']} ({a['chapter_id']}, Scene {a['scene_number']})")
        print(f"  Title: {a['title']}")
        print(f"  POV: {a['pov']}")
        print(f"  Characters: {', '.join(a['characters_involved'])}")
        conflict_preview = a['current_internal_conflict'][:100] if a['current_internal_conflict'] else 'None'
        print(f"  Current conflict: {conflict_preview}...")

if __name__ == '__main__':
    main()

File: scripts/analyze-scenes-151-200.ts
import * as fs from 'fs';
import * as path from 'path';

interface Scene {
  scene_number: number;
  chapter_id: string;
  title?: string;
  pov?: string;
  setup?: string;
  description?: string;
  focus?: string;
  internal_conflict?: string;
  internalConflict?: string;
  characters?: string[];
  [key: string]: any;
}

interface Chapter {
  chapter_id: string;
  scenes: Scene[];
  [key: string]: any;
}

interface Outline {
  chapters: Chapter[];
  [key: string]: any;
}

interface AnalysisResult {
  scene_index: number;
  chapter_id: string;
  scene_number: number;
  title: string;
  pov: string;
  current_internal_conflict: string;
  characters_involved: string[];
  proposed_enhanced_conflict: string;
}

// Read the outline file
const outlinePath = path.join(__dirname, '..', 'data', 'l_outline.json');
const outlineData: Outline = JSON.parse(fs.readFileSync(outlinePath, 'utf-8'));

// Extract all scenes with their global index
const allScenes: Array<Scene & { chapter_id: string; global_index: number }> = [];
let globalIndex = 0;

for (const chapter of outlineData.chapters) {
  if (chapter.scenes && Array.isArray(chapter.scenes)) {
    for (const scene of chapter.scenes) {
      globalIndex++;
      allScenes.push({
        ...scene,
        chapter_id: chapter.chapter_id,
        global_index: globalIndex
      });
    }
  }
}

console.log(`Total scenes found: ${allScenes.length}`);

// Extract scenes 151-200
const targetScenes = allScenes.filter(s => s.global_index >= 151 && s.global_index <= 200);

console.log(`Scenes 151-200: ${targetScenes.length} scenes`);

// Function to extract character names from text
function extractCharacters(scene: Scene & { chapter_id: string }): string[] {
  const characters = new Set<string>();

  // Add POV character
  if (scene.pov) {
    characters.add(scene.pov);
  }

  // Add from characters array if exists
  if (scene.characters && Array.isArray(scene.characters)) {
    scene.characters.forEach(c => characters.add(c));
  }

  // Common character names to look for
  const knownCharacters = [
    'Francisco', 'Zara', 'Roger', 'Maggie', 'Marcus', 'Kai', 'Elena',
    'Thomas', 'Sarah', 'David', 'Michael', 'Lisa', 'James', 'Amara',
    'Viktor', 'Isabella', 'Chen', 'Olivia', 'Nathan', 'Emma'
  ];

  // Search in text fields
  const textFields = [
    scene.setup || '',
    scene.description || '',
    scene.focus || '',
    scene.title || '',
    scene.internal_conflict || '',
    scene.internalConflict || ''
  ].join(' ');

  knownCharacters.forEach(name => {
    if (textFields.includes(name)) {
      characters.add(name);
    }
  });

  return Array.from(characters);
}

// Function to enhance internal conflict
function enhanceInternalConflict(scene: Scene & { chapter_id: string; global_index: number }): string {
  const currentConflict = scene.internal_conflict || scene.internalConflict || '';
  const pov = scene.pov || 'Unknown POV';
  const characters = extractCharacters(scene);

  // If there's already a detailed conflict, use it as base
  if (currentConflict && currentConflict.length > 100) {
    // Check if it already mentions other characters
    const hasOtherCharacters = characters.some(c =>
      c !== pov && currentConflict.includes(c)
    );

    if (hasOtherCharacters) {
      return currentConflict;
    }
  }

  // Build enhanced conflict based on available information
  const setup = scene.setup || '';
  const description = scene.description || '';
  const focus = scene.focus || '';
  const title = scene.title || '';

  let enhanced = currentConflict || '';

  // Add context from other fields if conflict is minimal
  if (enhanced.length < 50) {
    const contextParts = [];

    if (focus) contextParts.push(focus);
    if (description) contextParts.push(description);
    if (setup) contextParts.push(setup);

    if (contextParts.length > 0) {
      enhanced = contextParts.join(' ').substring(0, 300);
    }
  }

  // Add character interaction context if multiple characters are present
  if (characters.length > 1) {
    const otherCharacters = characters.filter(c => c !== pov);
    if (otherCharacters.length > 0 && !enhanced.includes(otherCharacters[0])) {
      enhanced += ` [Involves interactions with: ${otherCharacters.join(', ')}]`;
    }
  }

  return enhanced || `[Scene ${scene.global_index}: ${title}] - Internal conflict details needed`;
}

// Analyze each scene
const analysis: AnalysisResult[] = targetScenes.map(scene => {
  const characters = extractCharacters(scene);

  return {
    scene_index: scene.global_index,
    chapter_id: scene.chapter_id,
    scene_number: scene.scene_number,
    title: scene.title || 'Untitled',
    pov: scene.pov || 'Unknown',
    current_internal_conflict: scene.internal_conflict || scene.internalConflict || '',
    characters_involved: characters,
    proposed_enhanced_conflict: enhanceInternalConflict(scene)
  };
});

// Write the analysis to output file
const outputPath = path.join(__dirname, 'scenes-151-200-analysis.json');
fs.writeFileSync(outputPath, JSON.stringify(analysis, null, 2));

console.log(`\nAnalysis complete!`);
console.log(`Output written to: ${outputPath}`);
console.log(`\nSummary:`);
console.log(`- Total scenes analyzed: ${analysis.length}`);
console.log(`- Scenes with internal conflicts: ${analysis.filter(a => a.current_internal_conflict).length}`);
console.log(`- Scenes with multiple characters: ${analysis.filter(a => a.characters_involved.length > 1).length}`);

// Print sample
console.log(`\nSample scenes:`);
analysis.slice(0, 3).forEach(a => {
  console.log(`\nScene ${a.scene_index} (${a.chapter_id}, Scene ${a.scene_number})`);
  console.log(`  Title: ${a.title}`);
  console.log(`  POV: ${a.pov}`);
  console.log(`  Characters: ${a.characters_involved.join(', ')}`);
  console.log(`  Current conflict: ${a.current_internal_conflict.substring(0, 100)}...`);
});
